### ISSUE TITLE
Issues generating skin data

### ISSUE THREAD

**Issue by hYdos** [javagl/JglTF#94]
I'm generating a gltf skeleton and I have run into issues with data used. I am creating the skin [here](https://github.com/GenerationsMod/IMCT/blob/master/src/main/java/gg/generations/imct/scvi/flatbuffers/Titan/Model/SVModel.java#L312) and am parenting bones up [here](https://github.com/GenerationsMod/IMCT/blob/master/src/main/java/gg/generations/imct/scvi/flatbuffers/Titan/Model/SVModel.java#L107) which I have verified works however, I get errors from gltfValidator and no transforms on my bones. Any ideas whats going on?

Report attached in latest message

**Comment by hYdos** [comment]
I have solved the parenting issue. However, this also shows up which I missed. 
![image](https://github.com/javagl/JglTF/assets/25501603/1a14092d-055a-436e-9eaa-f9610189e66a)
Another issue is that the `DefaultNodeModel` matrices are not working. I have the model loaded in another format the expected way and compared it to the values in the matrix. Here is that comparison:
![image](https://github.com/javagl/JglTF/assets/25501603/c6450b8b-a509-4f13-ba05-ee9f7362f8cf)
This is what happens when i load it into blender though 
![image](https://github.com/javagl/JglTF/assets/25501603/4b59fcac-c1cf-453a-9b69-c15aa25bc77d)

**Comment by hYdos** [comment]
[out.zip](https://github.com/javagl/JglTF/files/12266748/out.zip)
Here is a copy of the model as well (zipped up because github moment)

**Comment by hYdos** [comment]
Another update: published the latest unreleased version and it fixed the override issue. The next zip will have images of the current blender state, the current model, and the gltf report.

[Info.zip](https://github.com/javagl/JglTF/files/12266976/Info.zip)

**Comment by javagl** [comment]
Skinning is complicated...

I'll have to allocate some time to take a closer look at the data and the relevant code paths. 

From a first quick look, regarding the `BUFFER_VIEW__TARGET_OVERRIDE` error: Which version of the library are you using? If you are using the version that is currently available in Maven Central, then it does not yet include the fix that was made in https://github.com/javagl/JglTF/commit/8d3aa2ee1e7b65e2dfffd3ebfca48bfae1966ee3 , and I think that this should resolve that particular issue.

(I don't know what the first screenshot is showing, for example, but maybe this will become clearer when I look at the data...)

**Comment by hYdos** [comment]
Yes i realised you had fixed that issue bit it was unreleased so I am now using a local build of JglTF to fix that error.

**Comment by hYdos** [comment]
We also found out you dont need inverse kinematics but JGltf is forcing us to have them

**Comment by javagl** [comment]
> We also found out you dont need inverse kinematics but JGltf is forcing us to have them

That's right. They should be optional. This should be fixed via https://github.com/javagl/JglTF/pull/95

Beyond that... I had a short look at this. The model is complex and contains many errors. The code that _generates_ the model is pretty complex as well, and I'll probably not be able to go though this (lacking time, having to check out the code, and... it would probably require the input model...).

However, the errors in the model are of three categories:

- `ACCESSOR_WEIGHTS_NON_NORMALIZED`: That's a bit surprising, because [this line in your code](https://github.com/GenerationsMod/IMCT/blob/8282e5b1b961b51729bd0f3368e2783c5f2b760a/src/main/java/gg/generations/imct/scvi/flatbuffers/Titan/Model/SVModel.java#L394) looks like it did some sort of normalization here...
- `ACCESSOR_INVALID_IBM`: You can either set all IBMs to be the identity matrix, or use the state from the PR that I linked to above, where the IBMs are _optional_ 
- `NODE_SKIN_NO_SCENE`: That's where something is structurally wrong.

Again, it's hard to point out where this error is coming from, but from a quick look, [this line](https://github.com/GenerationsMod/IMCT/blob/8282e5b1b961b51729bd0f3368e2783c5f2b760a/src/main/java/gg/generations/imct/scvi/flatbuffers/Titan/Model/SVModel.java#L342) seems to assign the same skin to many nodes, all with different meshes.... ? I'd have to spend much more time to understand what's going on there.

In the meantime, here is an example that creates an asset that is equivalent to the [`SimpleSkin`](https://github.com/KhronosGroup/glTF-Sample-Assets/tree/main/Models/SimpleSkin) sample asset:

```java
package de.javagl.jgltf.test.issue94;

import java.io.IOException;
import java.nio.FloatBuffer;
import java.nio.IntBuffer;
import java.nio.ShortBuffer;

import de.javagl.jgltf.model.AnimationModel.Interpolation;
import de.javagl.jgltf.model.GltfConstants;
import de.javagl.jgltf.model.creation.AccessorModels;
import de.javagl.jgltf.model.creation.GltfModelBuilder;
import de.javagl.jgltf.model.impl.DefaultAccessorModel;
import de.javagl.jgltf.model.impl.DefaultAnimationModel;
import de.javagl.jgltf.model.impl.DefaultAnimationModel.DefaultChannel;
import de.javagl.jgltf.model.impl.DefaultAnimationModel.DefaultSampler;
import de.javagl.jgltf.model.impl.DefaultGltfModel;
import de.javagl.jgltf.model.impl.DefaultMeshModel;
import de.javagl.jgltf.model.impl.DefaultMeshPrimitiveModel;
import de.javagl.jgltf.model.impl.DefaultNodeModel;
import de.javagl.jgltf.model.impl.DefaultSceneModel;
import de.javagl.jgltf.model.impl.DefaultSkinModel;
import de.javagl.jgltf.model.io.Buffers;
import de.javagl.jgltf.model.io.GltfModelWriter;

public class GenerateSimpleSkin
{
    public static void main(String[] args) throws IOException
    {
        DefaultAccessorModel indices = AccessorModels.createUnsignedShortScalar(
            IntBuffer.wrap(new int[] 
            { 
                0, 1, 3, 0, 3, 2, 2, 3, 5, 2, 5, 4, 
                4, 5, 7, 4, 7, 6, 6, 7, 9, 6, 9, 8 
            }
        ));
        DefaultAccessorModel positions = AccessorModels.createFloat3D(
            FloatBuffer.wrap(new float[] {
                -0.5f, 0, 0, 
                0.5f, 0, 0, 
                -0.5f, 0.5f, 0, 
                0.5f, 0.5f, 0, 
                -0.5f, 1, 0, 
                0.5f, 1, 0, 
                -0.5f, 1.5f, 0, 
                0.5f, 1.5f, 0, 
                -0.5f, 2, 0, 
                0.5f, 2, 0
            }
        ));
        
        DefaultAccessorModel joints = AccessorModels.create(
            GltfConstants.GL_UNSIGNED_SHORT, "VEC4", false, 
            Buffers.createByteBufferFrom(ShortBuffer.wrap(new short[] 
            {
                0, 0, 0, 0, 
                0, 0, 0, 0, 
                0, 1, 0, 0, 
                0, 1, 0, 0, 
                0, 1, 0, 0, 
                0, 1, 0, 0, 
                0, 1, 0, 0, 
                0, 1, 0, 0, 
                0, 1, 0, 0, 
                0, 1, 0, 0
            }
        )));
        
        DefaultAccessorModel weights = AccessorModels.createFloat4D(
            FloatBuffer.wrap(new float[] 
            {
                1, 0, 0, 0, 
                1, 0, 0, 0, 
                0.75f, 0.25f, 0, 0, 
                0.75f, 0.25f, 0, 0, 
                0.5f, 0.5f, 0, 0, 
                0.5f, 0.5f, 0, 0, 
                0.25f, 0.75f, 0, 0, 
                0.25f, 0.75f, 0, 0, 
                0, 1, 0, 0, 
                0, 1, 0, 0
            }
        ));
        
        DefaultAccessorModel inverseBindMatrices = AccessorModels.create(
            GltfConstants.GL_FLOAT,  "MAT4",  false, 
            Buffers.createByteBufferFrom(FloatBuffer.wrap(new float[] 
            {
                1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 
                1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, -1, 0, 1
            }
        )));
        
        
        DefaultAccessorModel animationTimes = AccessorModels.create(
            GltfConstants.GL_FLOAT, "SCALAR", false, 
            Buffers.createByteBufferFrom(FloatBuffer.wrap(new float[] 
            {
                0, 
                0.5f, 
                1, 
                1.5f, 
                2, 
                2.5f, 
                3, 
                3.5f, 
                4, 
                4.5f, 
                5, 
                5.5f
            }
        )));
        
        DefaultAccessorModel animationRotations = AccessorModels.createFloat4D(
            FloatBuffer.wrap(new float[] 
            {
                0, 0, 0, 1, 
                0, 0, 0.382999986410141f, 0.9240000247955322f, 
                0, 0, 0.7070000171661377f, 0.7070000171661377f, 
                0, 0, 0.7070000171661377f, 0.7070000171661377f, 
                0, 0, 0.382999986410141f, 0.9240000247955322f, 
                0, 0, 0, 1, 
                0, 0, 0, 1, 
                0, 0, -0.382999986410141f, 0.9240000247955322f, 
                0, 0, -0.7070000171661377f, 0.7070000171661377f, 
                0, 0, -0.7070000171661377f, 0.7070000171661377f, 
                0, 0, -0.382999986410141f, 0.9240000247955322f, 
                0, 0, 0, 1
            }
        ));
        
        
        // Meshes
        DefaultMeshModel mesh = new DefaultMeshModel();
        DefaultMeshPrimitiveModel meshPrimitive = 
            new DefaultMeshPrimitiveModel(4);
        meshPrimitive.setIndices(indices);
        meshPrimitive.putAttribute("POSITION", positions);
        meshPrimitive.putAttribute("JOINTS_0", joints);
        meshPrimitive.putAttribute("WEIGHTS_0", weights);
        mesh.addMeshPrimitiveModel(meshPrimitive);
        
        // Nodes
        DefaultNodeModel nodeWithMesh = new DefaultNodeModel();
        nodeWithMesh.addMeshModel(mesh);

        DefaultNodeModel jointNode0 = new DefaultNodeModel();
        DefaultNodeModel jointNode1 = new DefaultNodeModel();
        jointNode1.setTranslation(new float[] {0,1,0});
        
        // Skins
        DefaultSkinModel skin = new DefaultSkinModel();
        
        // NOTE: In this sample, these are the identity matrices,
        // and they will be OPTIONAL in the next release
        skin.setInverseBindMatrices(inverseBindMatrices);
        
        skin.addJoint(jointNode0);
        skin.addJoint(jointNode1);

        nodeWithMesh.setSkinModel(skin);
        jointNode0.addChild(jointNode1);

        // Animation
        DefaultAnimationModel animation = new DefaultAnimationModel();
        DefaultSampler animationSampler = new DefaultSampler(
            animationTimes, Interpolation.LINEAR, animationRotations);
        DefaultChannel animationChannel = new DefaultChannel(
            animationSampler, jointNode1, "rotation");
        animation.addChannel(animationChannel);
        
        // Scenes
        DefaultSceneModel scene = new DefaultSceneModel();
        scene.addNode(nodeWithMesh);
        scene.addNode(jointNode0);
        
        // Build the model
        GltfModelBuilder gltfModelBuilder = GltfModelBuilder.create();
        gltfModelBuilder.addSceneModel(scene);
        gltfModelBuilder.addAnimationModel(animation);
        DefaultGltfModel gltfModel = gltfModelBuilder.build();

        // Write the model, embedded, to the console
        GltfModelWriter w = new GltfModelWriter();
        w.writeEmbedded(gltfModel, System.out);
    } 
   
}
```

**Comment by javagl** [comment]
I think that this issue can be closed now. Some fixes that have been mentioned here should have been resolved with version 2.0.4. If there are further issues, they can be tracked separately.
